connectors/sources/graphql.py

```diff
@@ -0,0 +1,501 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""GraphQL source module responsible to fetch documents based on GraphQL Query."""
+import json
+import re
+from copy import deepcopy
+from functools import cached_property
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+from graphql import parse, visit
+from graphql.language.ast import VariableNode
+from graphql.language.visitor import Visitor
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import (
+    CancellableSleeps,
+    RetryStrategy,
+    hash_id,
+    iso_utc,
+    retryable,
+)
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+# Regular expression to validate the Base URL
+URL_REGEX = "^https?:\\/\\/(?:www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b(?:[-a-zA-Z0-9()@:%_\\+.~#?&\\/=]*)$"
+
+PING_QUERY = """
+{
+  __schema {
+    queryType {
+      name
+    }
+  }
+}
+"""
+
+
+class FieldVisitor(Visitor):
+    fields_dict = {}
+    variables_dict = {}
+
+    def enter_field(self, node, *args):
+        self.fields_dict[node.name.value] = []
+        self.variables_dict[node.name.value] = {}
+        if node.arguments:
+            for arg in node.arguments:
+                self.fields_dict[node.name.value].append(arg.name.value)
+                if isinstance(arg.value, VariableNode):
+                    self.variables_dict[node.name.value][
+                        arg.name.value
+                    ] = arg.value.name.value
+
+
+class UnauthorizedException(Exception):
+    pass
+
+
+class GraphQLClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.graphql_query = self.configuration["graphql_query"]
+        self.url = self.configuration["http_endpoint"]
+        self.graphql_object_list = self.configuration["graphql_object_list"]
+        self.http_method = self.configuration["http_method"]
+        self.authentication_method = self.configuration["authentication_method"]
+        self.pagination_technique = self.configuration["pagination_technique"]
+        self.pagination_key = self.configuration["pagination_key"]
+        self.valid_objects = []
+        self.variables = {}
+        self.headers = {}
```

Comment:
It will override [here](https://github.com/elastic/connectors/pull/2134/files#diff-1ac50fbc6d32ff7dfa228d41ad1b8945bd951532d30582faf148690d642ba397R451)